xapian-core/diversify/diversify.cc

```diff
@@ -144,6 +144,10 @@ DocumentSet
 Diversify::get_dmset(const MSet &mset)
 {
     LOGCALL(API, MSet, "Diversify::get_dmset", mset);
+    // Nothing to do
+    if (mset.empty())
+	throw InvalidArgumentError("Given mset should not be empty");
+
```

Comment:
I'd say if there's nothing to do then we should do nothing.  An empty `MSet` seems a totally reasonable answer to "give me a diversified version of this empty `MSet`".  It's already exactly as diverse as it can be.

If we throw an exception for an empty `MSet`, that forces the caller to check write code like:

```
Xapian::MSet diversified_mset;
if (mset.empty()) {
    diversified_mset = mset;
} else {
    diversified_mset = diversifier.get_dmset(mset);
}
```

In fact, I think the check in `get_dmset()` should be `if (mset.size() <= 2) return mset;` since diversification of a single entry `MSet` can only be a no-op, and swapping the entries in a two entry `MSet` won't make it any more diverse so the existing ordering from weighting or whatever should be respected.  Only once we get to three entries could reordering them conceivably improve diversity.

Reply:
Thanks for the tip. Agreed, diversification is fruitfully applicable to only three or more documents. I have added the necessary changes in a fixup commit.